Character spawning, by Chowdermcarthur

SubatomicTeddybear opened the pull request:
Fixes the Torch spawning only having cryo, I think. 
Poke @DeeUnderscore for more info (Or look at the code, you lazy bums)

chowdermcarthor committed:
Some sort of a spawning thing I dunno

DeeUnderscore commented:
I didn't write this. I just reabased someone else's code into one commit (and followed that with telling them to sort out the commit message and put their name on it).

[See more here](https://github.com/Baystation12/Baystation12/compare/dev...chowdermcarthor:dev).

SubatomicTeddybear commented:
Okay, so it's Chowder McArthur's code.  @chowdermcarthur

DeeUnderscore commented:
I changed the author on the commit. It's still a bit iffy, since it's not the author's commit message, but it's less iffy than signing somebody else's work with my name.

DeeUnderscore commented:
Looks like your thing passes CI, @chowdermcarthor.

(also, you misspelled it, @NobleCaos)

PsiOmegaDelta commented on code/game/jobs/job_controller.dm:
```diff
@@ -586,6 +587,13 @@ var/global/datum/controller/occupations/job_master
 		spawnpos = spawntypes[C.prefs.spawnpoint]
+	if(using_map)
+		var/datum/map/M = using_map
+		if(!(C.prefs.spawnpoint in M.allowed_spawns))
```
Could just be shortened to `if(!(C.prefs.spawnpoint in using_map.allowed_spawns))` without the prior if-check and cast.

PsiOmegaDelta replied:
Repeat for the other locations.

chowdermcarthor replied:
I don't use bay, this was a request for finalsong. I figured a sanity check would be better in case a map datum wasn't set.

GinjaNinja32 replied:
Map datums should always be set; it _not_ being set is an error anyway.

DeeUnderscore committed:
Tweak mapdata spawnpoint code

DeeUnderscore commented:
This is my life now.

I hope somebody actually did test this, because I sure haven't.

chowdermcarthor commented:
I tested it when it still had the check; it doesn't do anything on exodus(as is to be expected) but on torch, it warns as it's supposed to, and spawns people at a cryo spawn if they don't have cryo set.

DeeUnderscore commented:
Awesome.

We use `using_map` all over the place, so if for some weird reason it's missing, crashing loudly is a reasonable outcome.

GinjaNinja32 commented on code/game/jobs/job_controller.dm:
```diff
@@ -586,6 +587,13 @@ var/global/datum/controller/occupations/job_master
+		if(!(C.prefs.spawnpoint in M.allowed_spawns))
+			if(H)
+				H << "Your chosen spawnpoint ([spawnpos.display_name]) is unavailable for the current map. Spawning you at one of the enabled spawn points instead."
+				H.forceMove(pick(latejoin_cryo))//This for torch really
```
Should likely either just be `latejoin`, or a map-definable list with the one now on line 612 changed to the same.

chowdermcarthor replied:
latejoin, as it turns out, is what the arrivals shuttle uses. I didn't feel like making a new one for that, or making a list, and for now as cryo is the one that'll be used, I just used that.

GinjaNinja32 replied:
Spawnpoints that are unavailable for the current _job_ will revert to `latejoin`; if that's not correct for at least one map, line 612 needs changing to a map-specific system, which will mean this line has a better option too.

DeeUnderscore committed:
Replace LateSpawn()

LateSpawn() was used both for picking a place to spawn and returning it,
and actually moving mobs. It has been replaced with get_spawnpoint_for(),
which picks a place to spawn and returns it. Moving mobs is then handled
by calling procs.

DeeUnderscore committed:
Set allowed_spawns for the map Example

DeeUnderscore committed:
Add test for incorrect entries in allowed_spawns

DeeUnderscore committed:
Add allowed_spawns to overmap_example

DeeUnderscore commented:
I made some changes here, so please re-review.
- `LateSpawn()` was replaced with `get_spawnpoint_for()`. `LateSpawn()` would either return a spawn turf, or move a player to a spawn turf and return a spawn message, depending on arguments. Instead of that `get_spawnpoint_for()` just returns the appropriate spawnpoint, and it's up to the calling procs to actually do the moving.
- If a spawn pref is unset, or incorrect for the job, the fallback is to go through all the allowed spawns for map and select the first one correct for the given job. If that fails (which is an error), a spawn point is picked at random from all the available ones for the map instead. 
- A test checks if all spawn points in `allowed_spawns` have turfs and are thus actually spawnable. This was no the case for Example, and the test should catch any such situations in the future.
